### PR TITLE
fix/chain_head: Ensure correct events for finalized branch

### PR DESCRIPTION
**Pull request by lexnv**
Issue https://github.com/paritytech/substrate/issues/11321 has detected a few flaky tests coming form the chainHead.

### Error
The error encountered (extracted from [here](https://github.com/paritytech/substrate/pull/13631#discussion_r1140365085)):

```bash
--- STDERR:              sc-rpc-spec-v2 chain_head::tests::follow_forks_pruned_block ---
thread 'chain_head::tests::follow_forks_pruned_block' panicked at 'assertion failed: `(left == right)`
  left: `Finalized(Finalized { finalized_block_hashes: ["0x0af997b079e96b7f333bf2fe71cb914ad3eb49355923fd3db96c0cc91243bd7a"], pruned_block_hashes: [] })`,
 right: `BestBlockChanged(BestBlockChanged { best_block_hash: "0x0af997b079e96b7f333bf2fe71cb914ad3eb49355923fd3db96c0cc91243bd7a" })
```

### Investigation
The issue happens on the following code path:
- the finalized event is triggered before the new_block event
- finalized branch is responsible for generating NewBlock + BestBlock event
https://github.com/paritytech/substrate/blob/12c323596eb3b44ea4a9fcd9a96ea03d2d8563e9/client/rpc-spec-v2/src/chain_head/chain_head_follow.rs#L335-L341

- the `best_block_cache` is set, however the intent was just to verify if we have encountered a gap in 
https://github.com/paritytech/substrate/blob/0506da00297757babe63b7bb62e487c657066117/client/rpc-spec-v2/src/chain_head/chain_head_follow.rs#L394-L404
- the `generate_import_events()` checks if the best block was already encountered to not regenerate the event. 
https://github.com/paritytech/substrate/blob/0506da00297757babe63b7bb62e487c657066117/client/rpc-spec-v2/src/chain_head/chain_head_follow.rs#L295-L297

### Fix

Remove populating the `best_block_cache` on the finalized branch to allow `generate_import_events()` to properly generate the `BestBlock` event for the latest finalized hash. 

### Testing Done
- ran `cargo test` in a loop with `stress-ng --cpu 6 --io 4` for 239 times on my machine. Although it gives us some confidence, I've only encountered the `Finalized` event arriving before `NewBlock` just a few times in a more intensive test while introducing the code.

### Notes
- [x] Renable chainHead tests if https://github.com/paritytech/substrate/pull/13631 is merged first

// CC @paritytech/tools-team

### CODE REVIEW

File: client/rpc-spec-v2/src/chain_head/chain_head_follow.rs [hunk 1]
```diff
@@ -373,31 +373,26 @@ where
...
+
+				// A descendent of the finalized block was already reported
+				// before the `NewBlock` event containing the finalized block
+				// is reported.
+				if ancestor.hash == *last_finalized {
+					return Err(SubscriptionManagementError::Custom(
+						"A descendent of the finalized block was already reported".into(),
+					))
+				}
+			}
```
Comment (bkchr) [inline client/rpc-spec-v2/src/chain_head/chain_head_follow.rs:395]: I don't get this piece of code? How do you know that there wasn't any new best block event for the last finalized block?

And why do you have this as part of of the for loop while it doesn't depend on any from the loop?

Comment (lexnv) [reply client/rpc-spec-v2/src/chain_head/chain_head_follow.rs:395]: Hey

> How do you know that there wasn't any new best block event for the last finalized block?

We depend on the `pin_block` which returns `true` if this is the first time we "pin" the block and `false` if the block was previously pinned. We call `pin_block` twice for each block: from the import and finalized branches.

https://github.com/paritytech/substrate/blob/df6a43822dd150418832279ea1687d508d0f7e93/client/rpc-spec-v2/src/chain_head/chain_head_follow.rs#L364-L369


> And why do you have this as part of of the for loop while it doesn't depend on any from the loop?

For any blocks that we have not seen yet we want to generate the NewBlock, except for the last block for which we want to generate the `NewBlock` + `BestBlock` events. For the last block, I've added the extra check to make sure we catch the following case:

```bash
[Block 0] -  [Block 1] - [Block 2] -  ...
                          ^^^ T0: Block import branch reports block 2 for the first time 

^^^^^^^^^^^^^^^ T1:  Finalized branch reports 0 and 1 for the first time
```

We could enter this path only if the BlockImport notification is not generated for 0, 1; but somehow we receive the  BlockImport for block 2. And at the same time we get Finalized for 0 and 1 (blocks that we have not been reported by BlockImport). 

This check may be an overcautious one, let me know what you think of it, thanks!

**Review by bkchr — APPROVED** [review]

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2565605

File: client/rpc-spec-v2/src/chain_head/chain_head_follow.rs [hunk 2]
```diff
@@ -350,58 +344,47 @@ where
 		};
 
 		// Find the parent hash.
-		let Some(first_number) = self.client.number(*first_hash)? else {
-			return Err(SubscriptionManagementError::BlockNumberAbsent)
-		};
-		let Some(parent) = self.client.hash(first_number.saturating_sub(One::one()))? else {
+		let Some(first_header) = self.client.header(*first_hash)? else {
 			return Err(SubscriptionManagementError::BlockHashAbsent)
```
Comment (skunert) [inline client/rpc-spec-v2/src/chain_head/chain_head_follow.rs:348]: super nit: This sounds like the block hash is absent, but its more `HeaderNotFound` or something.

File: client/rpc-spec-v2/src/chain_head/chain_head_follow.rs [hunk 3]
```diff
@@ -350,58 +344,47 @@ where
 		};
 
 		// Find the parent hash.
```
Comment (skunert) [inline client/rpc-spec-v2/src/chain_head/chain_head_follow.rs:346]:
```suggestion
		// Find the parent header.
```

**Review by skunert — APPROVED** [review]

File: client/rpc-spec-v2/src/chain_head/chain_head_follow.rs (hunk 2 above)
Comment (lexnv) [reply client/rpc-spec-v2/src/chain_head/chain_head_follow.rs]: That makes sense! I've changed it to `BlockHeaderAbsent` to make it explicit that we are missing the header from the db, thanks!

**Pull request merged**